### DISCUSSION TITLE
Use of the normal users table vs a dedicated one?

### DISCUSSION THREAD

**Discussion by edgrosvenor** [filamentphp/filament#30] [Help] [answered]
In most cases, I don't want to have a dedicated users table for administrators. I already have roles and something akin to an is_admin column in my regular old users table. Having skimmed the code it looks like just using the standard Laravel users table is a trivial matter and I'm sure it'll take me next to no time to get it set up. I guess I'm just wondering whether it would be worth making the User model a configuration setting so that it's truly trivial for anyone that doesn't want a separate table of users. Or, at the very least, would it make sense to add instructions to the documentation on how to accomplish it.

Either way, great looking package. I plan to pull it into something I'm working on right now (and hack away until my existing admin users can use it) and can't wait to see it in action!

**Comment by Wulfheart** [comment]
Definitely second this. In my use-case I would like to call a method ``isAdmin()`` on my user model. Maybe this can be achieved through a trait or an interface?

**Comment by danharrin** [comment]
Hi! Thanks for the suggestion.

I'm open to making the users table configurable. However, and I probably didn't make this clear enough in the docs, the `is_admin` property is used to differentiate between normal Filament users and Filament admins, who have "root" access everywhere and can manage other users.

I could implement checks like `isFilamentUser()` and `isFilamentAdmin()` if you wish?

**Comment by iksaku** [comment]
Another way to tackle this, as mentioned in Discord by @ryangjchandler, is to migrate to use [Laravel Gates and Policies](https://laravel.com/docs/8.x/authorization) instead of using filament specific DB tables and models. This is the way that Nova handles their CMS access.

If implemented, it will mean a breaking change, and maybe most of us could just accept it as a `v1.1` or `v1.2`.

Any other ideas?

**Reply by gavinhewitt** [reply]
@iksaku Which Discord is that? Is there a dedicated Filament Discord?

**Reply by iksaku** [reply]
We're on the Livewire discord, under the #filament channel

**Comment by danharrin** [comment]
Hey @edgrosvenor @Wulfheart @iksaku. I've opened #72 with support for this feature. Would be great if you could review, thanks!

**Reply by lewislarsen** [reply]
Awesome, Dan!

Does the PR maintain separate sessions and require the user to re-authenticate to access Filament? 

Gave Filament a little go earlier today, good stuff. :)

**Reply by ryangjchandler** [reply]
@LewisLarsen If you change the guard that Filament uses, only a single session will be maintained (assuming the guard is the same guard used for your custom User model).

**Reply by edgrosvenor** [reply]
@danharrin Having read through the PR, it looks like it addresses all of my needs. I'll have to give a little thought to the admin vs regular filament user thing because that framework doesn't fit neatly within the context of what I'm working on. But that's a deficit in my own mental model, not your architecture. I can definitely do what I need to do with what's in this branch.

Thanks for your hard work on this package. It's the Game Changer of 2021 so far and sets a pretty high bar for the rest of the year.

**Reply by ryangjchandler** [reply]
@edgrosvenor The idea of the "is Filament user" concept is just to prevent people from accessing Filament full-stop, i.e. normal customers.

The `is_admin` check, out of the box atleast, just prevents the user from accessing the User resource.

**Reply by danharrin** [reply]
And you can omit the admin column and all users will be able to manage other users.

**Reply by edgrosvenor** [reply]
In my case, though, I have 3 tiers of users with ownership over some data. Since eliminating the admin column altogether is possible, I'll just do that and rely on gates. If I get hung up on it I'll let you know but it all feels pretty straightforward now. Thank you!

**Reply by ryangjchandler** [reply]
@edgrosvenor Out of interest, how are you going to rely on gates? There's no support for policies / custom auth checks besides the `hasFilamentRole` method on the user.

**Reply by danharrin** [reply]
I think policy support would be a good next step.

**Reply by ryangjchandler** [reply]
> I think policy support would be a good next step.

Pair again this weekend? 😅

**Reply by danharrin** [reply]
Sure!

**Reply by iksaku** [reply]
#72 is a good step forward, but the need for Policies is what I'd rather use, that way I can let users access different resources based on special roles/permissions.

Still, I can see so much great stuff coming from you all!

**Reply by ryangjchandler** [reply]
Yeah, for policies to work though, you'd need custom user support otherwise you'd need multiple policies for each user type.

**Reply by edgrosvenor** [reply]
@ryangjchandler I actually hadn't thought that whole gate thing through. It was one of those "I'll figure it out when I get there" statements. Turns out the answer was "say I'm going to do it and hopefully someone will help make it possible." I'm amazed at how quickly you guys are moving on this.

**Reply by ryangjchandler** [reply]
@edgrosvenor Me and Dan are going to pair on the gates & policies this weekend at some point so hopefully we'll have an answer for you soon!

**Comment by danharrin** [comment, marked as answer]
Supported from v1.3.0.

**Reply by mallardduck** [reply]
This is pretty easy to pull off. I've got a local dev space I'm playing around with filament in where I've made the User's model extend Filament's user model. Then I override the `$table` property to `users` and make sure my migration matched the default filament prefixed one.

I know this is a bit different than documented - but in my case it feels easier as I'm doing minimal customization.
